### PR TITLE
WIP: MetadataStore module ("Channel 2.0")

### PR DESCRIPTION
**Pull request by ichorid**
This module is the base of our upcoming Channel2.0 subsystem. Currently, it provides 3 features:
* PonyORM-based objects/DB bindings for Torrents and Channels metadata
* Serialization-deserialization of Torrent and Channels metadata (XDRLib)
* Torrent-based dissemination of Channels (new-style Channel can be serialized to disk in the form of a torrent)

At the moment, the module is integrated with _some_ of the REST endpoints. It is done in a dual-stack manner (i.e. search endpoint returns concatenated search results over _both_ new- and old-style channels).
By default, it is possible to load the .mdblob file containing channel metadata entry from disk. Tribler should start to download the corresponding torrent automatically into the correct "channels dir", and add scan the dir for binary metadata blobs when the download finishes. This is the intended usage mode for first experimental versions - just download channels by hand.
To enable the ability to create the new-style channels, the user _must_ tick the corresponding checkbox in the GUI "preferences" dialog box. If the option is enabled, new-style channel will be created from "My channel" dialog.

**TODO and problems**
* **First**, as @xoriole pointed out, we still don't know how to seed unique torrents from behind the NAT. One solution can be using hidden seeding (if it works).
* **Second**, there is a LOT of stuff to be changed in the remaining REST endpoints to support new-style channel addresses (which are basically hex-encoded IPv8 public keys).
* **Third** the saved "personal channel" is not seeded by default. 
* **Fourth** current channel "torrent" form uses "single MD entry per file" scheme. It is inefficient for very large channels, but very convenient for testing purposes.
* And add more unit tests, and state checks for channel downloads, and move stuff around, and refactor unit tests, etc, etc...

So, the draft plan is: 
1. Fix the remaining REST endpoints (and add tests for them).
2. Add tracker field to the TorrentMD object and serialized form.
3. Get the channel torrents seeding working.
4. Start squashing all MD entries from the same channel update into a single file.
5. Work through TODOs and FIXMEs in the code.

**Design notes**
* Some methods are "dangling" and not bound to objects. This is because there is still no fitting objects to attach them to. This objects should emerge eventually.
* We have an option to make serialization procedure completely automatical based on class structure.
* The class structure is designed to be general, but not *too* general.
* `serialize` method **changes input dict by design**. It is discussable, though.

### CODE REVIEW

File: Tribler/Test/Core/Modules/MetadataStore/tools.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+# FIXME global vars! shadowing!
+crypto = ECCrypto()
+key = crypto.generate_key('curve25519')
+public_key = key.pub().key_to_bin()
+
+
+def are_dir_trees_equal(dir1, dir2):
+    """
+    Shameless copypaste from Stackexchange.
```
Comment (qstokkink) [inline Tribler/Test/Core/Modules/MetadataStore/tools.py]: Source please.

**Comment by qstokkink** [comment]
Does anyone volunteer to install pony on all the testing machines? 😃

**Comment by ichorid** [comment]
Or, better yet, add Pony to dependencies list and write automatical dependency install script...

File: Tribler/Test/Core/Modules/MetadataStore/tools.py (hunk 1 above)
Comment (ichorid) [reply Tribler/Test/Core/Modules/MetadataStore/tools.py]: https://stackoverflow.com/questions/4187564/recursive-dircmp-compare-two-directories-to-ensure-they-have-the-same-files-and

**Comment by qstokkink** [comment]
@ichorid #1768, #3770.. our dependencies are quite a pain: we have some which can be `pip`'d, but others that are incredibly broken on their latest versions.

**Comment by xoriole** [comment]
retest this please

**Comment by qstokkink** [comment]
Hmm.. two segfaults in the Linux tests.

**Comment by ichorid** [comment]
@qstokkink , you're using Pony 0.7.6rc1. Its not stable yet, this could probably be the case. Try 0.7.5.

**Comment by qstokkink** [comment]
@ichorid I updated from 0.7.5 because I was getting that error there as well.

**Comment by ichorid** [comment]
@qstokkink , oops...

**Comment by qstokkink** [comment]
@ichorid I'm guessing this is a cascaded failure due to `TestDownloadChannel.test_download_channel` and if that test fails it does not unlock the `db_session` or something.

**Comment by qstokkink** [comment]
@ichorid That was the case. I fixed it as follows:

```patch
diff --git a/Tribler/Test/test_as_server.py b/Tribler/Test/test_as_server.py
index 13a28ec..71cd421 100644
--- a/Tribler/Test/test_as_server.py
+++ b/Tribler/Test/test_as_server.py
@@ -126,6 +126,13 @@ class AbstractServer(BaseTestCase):
                 self._logger.error(">     %s" % dc)
                 dc.cancel()
 
+        from pony.orm.core import local
+        if local.db_context_counter > 0:
+            self._logger.error("Leftover pony db sessions found!")
+            from pony.orm import db_session
+            for _ in range(local.db_context_counter):
+                db_session.__exit__()
+
         has_network_selectables = False
         for item in reactor.getReaders() + reactor.getWriters():
             if isinstance(item, HTTPChannel) or isinstance(item, Client):
```

This cleans up any leaking `db_session`s.

**Comment by ichorid** [comment]
@qstokkink , you're already better at Pony handling than me :wink:

**Comment by qstokkink** [comment]
@ichorid also, please use this nasty, nasty hack to make the download test more consistent:

```patch
diff --git a/Tribler/Core/Modules/MetadataStore/channels.py b/Tribler/Core/Modules/MetadataStore/channels.py
index d0b8a1e..823f9b9 100644
--- a/Tribler/Core/Modules/MetadataStore/channels.py
+++ b/Tribler/Core/Modules/MetadataStore/channels.py
@@ -140,6 +140,7 @@ def download_channel(session, infohash, title):
     dcfg.set_dest_dir(session.channels_dir)
     tdef = TorrentDefNoMetainfo(infohash=str(infohash), name=title)
     download = session.start_download_from_tdef(tdef, dcfg)
+    session.lm.ltmgr.dht_ready = True
 
     def err(f):
         print "we got an exception: %s" % (f.getTraceback(),)
```

**Comment by ichorid** [comment]
``` diff
+    session.lm.ltmgr.dht_ready = True
```
@qstokkink , do we really have to put it into the download method itself? We only want DHT to be disabled for unit tests.

**Comment by qstokkink** [comment]
@ichorid you could also generally enable that, but I don't know if it will screw with the other tests. It's safer to only do that for these tests.

**Comment by ichorid** [comment]
@qstokkink , the test times out if I add this line in the test itself.

**Comment by ichorid** [comment]
Seems like some of the tests added in this PR fail on timeout on test machines. This is clearly not the case on my development machines, though...
@qstokkink , is there some way to force tests to run sequentially for this PR, so we could see exactly which one causes the failure?

**Comment by qstokkink** [comment]
@ichorid not without making an entirely custom Jenkins job (if it was any error other than a segfault you would be able to see this per test though).

**Comment by qstokkink** [comment]
Hmm it seems there is something VERY wrong with the linux machine: `INFO:OneShotProcessProtocol:[run_nosetests_for_je...] ERR: ERROR: ld.so: object '/usr/lib/libtcmalloc_minimal.so.4' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.`.
My python-for-android builder is also quite dead. Something funky with `ld.so`.

**Comment by ichorid** [comment]
@xoriole , could you comment on what happens with the Linux test machine?

**Comment by synctext** [comment]
Solid progress..
About NATs. Lets just assume for 2018 that the people creating channels are connectable. Would be great to inform users of their connectability and max. download speed.

**Comment by ichorid** [comment]
retest this please

**Comment by qstokkink** [comment]
@ichorid how far along is this PR?

**Comment by ichorid** [comment]
@qstokkink, I'll get back to work this monday and fix test timeouts.

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
We have a race condition that causes exceptions on interpreter teardown.
To trigger it, 3 conditions must be fulfilled:
1. Run the code with `nosetests`. Using `trial` or just running Tribler never triggers it.
2. Complete real download with Libtorrent. Apparently, our codebase does not include real Libtorrent-based download tests. The only such test (`test_download.py`) was introduced in this pull request.
3. Bind DB file with PonyORM.

Then, on my PC it is triggered by `test_download.py` with 20% probability:
```
Exception in thread Twisted (most likely raised during interpreter shutdown):test_download_channel
```

Most probably its caused by Pony using `pysqlite`, [which is known to be worse than APSW in regards to threads and locking](https://www.reddit.com/r/Python/comments/yx20w/the_sqlite_lock_timeout_nightmare/).
I've sent the question to Pony's author regarding this behavior and a possible switch to APSW.

**Comment by ichorid** [comment]
Adding some `sleep` on `tearDown` of `test_download` fixes the exception problem, *but only if `test_download` was the last test run by `nosetests`*. Obviously, the problem has something to do with how `nosetests` handles interpreter shutdown.

**Comment by ichorid** [comment]
The last build has `test_downloads.py` removed, and it still crashes. The funniest thing is, I can't see any problems in "archived artifacts" logs.
Help on this is really appreciated...

**Comment by devos50** [comment]
I can help you tomorrow if you want

**Comment by qstokkink** [comment]
https://github.com/Tribler/gumby/blob/ff7f7996a8b99c171f6367d9c7280eb43458dcc0/scripts/run_nosetests_for_jenkins.sh#L183

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
Windows error:

```python
Traceback (most recent call last):
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\pyipv8\ipv8\util.py", line 14, in helper
    return blockingCallFromThread(reactor, func, *args, **kargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\pyipv8\ipv8\util.py", line 41, in blockingCallFromThread
    result.raiseException()
  File "<string>", line 2, in raiseException
OperationalError: no such module: FTS4
```

**Comment by qstokkink** [comment]
We added fts4 to the Windows tester. There are remaining bugs though:

```python
======================================================================
ERROR: test_commit_to_torrent_add_torrents (Tribler.Test.Core.Modules.MetadataStore.test_channels_md.TestChannelMD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\test_as_server.py", line 58, in check
    result = fun(*argv, **kwargs)
  File "<auto generated wrapper of test_commit_to_torrent_add_torrents() function>", line 2, in test_commit_to_torrent_add_torrents
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\Core\Modules\MetadataStore\test_channels_md.py", line 126, in test_commit_to_torrent_add_torrents
    chan.get_dirname()))
  File "<auto generated wrapper of process_channel_dir() function>", line 2, in process_channel_dir
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\channels.py", line 122, in process_channel_dir
    load_blob(db, os.path.join(dirname, filename))
  File "<auto generated wrapper of load_blob() function>", line 2, in load_blob
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\channels.py", line 164, in load_blob
    gsp = deserialize_metadata_gossip(f.read())
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\serialization.py", line 92, in deserialize_metadata_gossip
    md["timestamp"] = float2time(u.unpack_double())
  File "c:\python27\lib\xdrlib.py", line 207, in unpack_double
    raise EOFError
EOFError: 
-------------------- >> begin captured logging << --------------------
```

```python
======================================================================
ERROR: test_process_channel_dir (Tribler.Test.Core.Modules.MetadataStore.test_channels_md.TestChannelMD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\test_as_server.py", line 58, in check
    result = fun(*argv, **kwargs)
  File "<auto generated wrapper of test_process_channel_dir() function>", line 2, in test_process_channel_dir
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\Core\Modules\MetadataStore\test_channels_md.py", line 164, in test_process_channel_dir
    process_channel_dir(self.session.mds, sample_channel_dir)
  File "<auto generated wrapper of process_channel_dir() function>", line 2, in process_channel_dir
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\channels.py", line 122, in process_channel_dir
    load_blob(db, os.path.join(dirname, filename))
  File "<auto generated wrapper of load_blob() function>", line 2, in load_blob
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\channels.py", line 164, in load_blob
    gsp = deserialize_metadata_gossip(f.read())
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Core\Modules\MetadataStore\serialization.py", line 110, in deserialize_metadata_gossip
    md["signature"] = u.unpack_opaque()
  File "c:\python27\lib\xdrlib.py", line 224, in unpack_string
    return self.unpack_fstring(n)
  File "c:\python27\lib\xdrlib.py", line 216, in unpack_fstring
    raise EOFError
EOFError: 
-------------------- >> begin captured logging << --------------------
```

```python
======================================================================
FAIL: test_garbage_collect (Tribler.Test.Core.Modules.MetadataStore.test_channels_md.TestChannelMD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\test_as_server.py", line 58, in check
    result = fun(*argv, **kwargs)
  File "<auto generated wrapper of test_garbage_collect() function>", line 2, in test_garbage_collect
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\Core\Modules\MetadataStore\test_channels_md.py", line 80, in test_garbage_collect
    self.assertNotIn(del1, del_entries)
AssertionError: DeletedMD[1] unexpectedly found in [DeletedMD[1], DeletedMD[3]]
-------------------- >> begin captured logging << --------------------
```

```python
======================================================================
FAIL: test_list_new_entries (Tribler.Test.Core.Modules.MetadataStore.test_channels_md.TestChannelMD)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\test_as_server.py", line 58, in check
    result = fun(*argv, **kwargs)
  File "<auto generated wrapper of test_list_new_entries() function>", line 2, in test_list_new_entries
  File "c:\python27\lib\site-packages\pony\orm\core.py", line 466, in new_func
    result = func(*args, **kwargs)
  File "C:\workspace\jenkins2\workspace\GH_Tribler_PR_tests_win64\tribler\Tribler\Test\Core\Modules\MetadataStore\test_channels_md.py", line 60, in test_list_new_entries
    self.assertIn(md2, channel1.newer_entries())
AssertionError: TorrentMD[3] not found in []
-------------------- >> begin captured logging << --------------------
```

**Comment by devos50** [comment]
Also, please rebase this PR on the `devel` branch since it will be a feature of Tribler 7.2.

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
We have some SSL module problems on our test machines?

Windows:
```python
:0: UserWarning: You do not have a working installation of the service_identity module: 'cannot import name opentype'.  Please install it from <https://pypi.python.org/pypi/service_identity> and make sure all of its dependencies are satisfied.  Without the service_identity module, Twisted can perform only rudimentary TLS client hostname verification.  Many valid certificate/hostname mappings may be rejected.
```

MacOS:
```
Failure instance: Traceback (failure with no frames): <class 'twisted.web._newclient.ResponseNeverReceived'>: [<twisted.python.failure.Failure OpenSSL.SSL.Error: [('SSL routines', 'SSL23_GET_SERVER_HELLO', 'tlsv1 alert protocol version')]>]
]
```

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
Mental note, error on non-ascii torrent filename:
```python
Traceback (most recent call last):
  File "/home/vader/.local/lib/python2.7/site-packages/twisted/web/server.py", line 198, in process
    self.render(resrc)
  File "/home/vader/.local/lib/python2.7/site-packages/twisted/web/server.py", line 258, in render
    body = resrc.render(self)
  File "/home/vader/.local/lib/python2.7/site-packages/twisted/web/resource.py", line 250, in render
    return m(request)
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/restapi/channels/channels_torrents_endpoint.py", line 166, in render_PUT
    self.session.mds.TorrentMD.from_tdef(key, torrent_def, extra_info)
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/MetadataStore/torrents.py", line 34, in from_tdef
    "tc_pointer": 0})
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/MetadataStore/base.py", line 43, in from_dict
    md.sign(key)
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/MetadataStore/base.py", line 36, in sign
    serialize_metadata_gossip(md_dict, key)
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/MetadataStore/serialization.py", line 65, in serialize_metadata_gossip
    p.pack_string(md["title"])
  File "/usr/lib/python2.7/xdrlib.py", line 111, in pack_string
    self.pack_fstring(n, s)
  File "/usr/lib/python2.7/xdrlib.py", line 104, in pack_fstring
    self.__buf.write(data)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 17-23: ordinal not in range(128)
```

**Comment by ichorid** [comment]
Another mental note: 
```python
ERROR   1535012905.11             sqlitecachedb:293  (SQLiteCacheDB)  cachedb: ===MainThread===
SQL Type: <type 'unicode'>
-----
UPDATE Torrent SET num_files=?,length=?,name=? WHERE infohash='ZR9MOhGqyLMjstsrsU0sgII/DJQ='
-----
[115, 983536841, '\xd0\x92\xd0\xb8\xd0\xba\xd1\x82\xd0\xbe\xd1\x80 \xd0\x90\xd1\x80\xd0\xb3\xd0\xbe\xd0\xbd\xd0\xbe\xd0\xb2 Project']
======
Traceback (most recent call last):
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/sqlitecachedb.py", line 284, in execute
    return cur.execute(sql, args)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xd0 in position 0: ordinal not in range(128)
Unhandled error in Deferred:

ERROR   1535012905.15            tribler_window:90   (root)  Traceback (most recent call last):
  File "/home/vader/my_SRC/TRIBLER/tribler_ichorid/TriblerGUI/event_request_manager.py", line 127, in on_read_data
    raise RuntimeError(json_dict["event"]["text"])
RuntimeError: [Failure instance: Traceback: <type 'exceptions.UnicodeDecodeError'>: 'ascii' codec can't decode byte 0xd0 in position 0: ordinal not in range(128)
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Libtorrent/LibtorrentMgr.py:584:_task_process_alerts
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Libtorrent/LibtorrentMgr.py:532:got_metainfo
/home/vader/.local/lib/python2.7/site-packages/twisted/internet/defer.py:460:callback
/home/vader/.local/lib/python2.7/site-packages/twisted/internet/defer.py:568:_startRunCallbacks
--- <exception caught here> ---
/home/vader/.local/lib/python2.7/site-packages/twisted/internet/defer.py:654:_runCallbacks
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/Modules/restapi/torrentinfo_endpoint.py:71:on_got_metainfo
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/SqliteCacheDBHandler.py:533:update_torrent_with_metainfo
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/SqliteCacheDBHandler.py:513:updateTorrent
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/sqlitecachedb.py:370:update
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/sqlitecachedb.py:326:execute_write
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/pyipv8/ipv8/util.py:14:helper
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/pyipv8/ipv8/util.py:26:blockingCallFromThread
/home/vader/my_SRC/TRIBLER/tribler_ichorid/Tribler/Core/CacheDB/sqlitecachedb.py:295:execute
```

**Comment by ichorid** [comment]
1st test works, 2nd test works, 3rd test crashes **if `:memory:` database is forced in Session itself**.
However, the regular tribler always works OK.
```python

from Tribler.Core.Modules.MetadataStore import base
from pony import orm
from pony.orm import db_session
from Tribler.Test.test_as_server import TestAsServer


def test1():
    db = base.start_orm(":memory:", True)

    with db_session:
        db.SignedGossip()
        print orm.select(g for g in db.SignedGossip)[:]
    with db_session:
        db.SignedGossip()
        print orm.select(g for g in db.SignedGossip)[:]

class some_test_tribler_class(TestAsServer):
    def test2(self):
        db = base.start_orm(":memory:", True)

        with db_session:
            db.SignedGossip()
            print orm.select(g for g in db.SignedGossip)[:]
        with db_session:
            db.SignedGossip()
            print orm.select(g for g in db.SignedGossip)[:]

class some_test_tribler_class(TestAsServer):
    def test2(self):
        # has to change Session to use :memory: based sqlite db to crash this
        db = self.session.mds

        with db_session:
            db.SignedGossip()
            print orm.select(g for g in db.SignedGossip)[:]
        with db_session:
            db.SignedGossip()
            print orm.select(g for g in db.SignedGossip)[:]
```

However, **`trial`** works OK with this!
WTF??!! :face_with_head_bandage:

**Comment by ichorid** [comment]
OK, seems like we've hit [this bug](https://bugs.python.org/issue1856).

Test code:
```python
#!/usr/bin/python2.4 -u

"""Tickle python threading bugs."""

import os
import random
import sys
import time
import threading

thread_has_run = set()

class RandomLifeThread(threading.Thread):
  """Loop for a while sleeping random tiny amounts and doing some I/O."""
  def run(self):
    while True:
      zero_f = open('/dev/zero', 'r')
      blank = zero_f.read(200)
      null_f = open('/dev/null', 'w')
      null_f.write(blank)
      zero_f.close()
      time.sleep(random.random()/199)
      null_f.close()
      thread_has_run.add(self.name)

def main(_):
  count = 0
  for _ in xrange(40):
    new_thread = RandomLifeThread()
    new_thread.setDaemon(True)
    new_thread.name = 'Thread-%010x' % count
    new_thread.setName(new_thread.name)
    new_thread.start()
    count += 1
  #sys.setcheckinterval(random.randint(0, 100))
  while len(thread_has_run) < count:
    time.sleep(0.001)
  sys.exit(0)

def some_test():
    main(sys.argv)
```

**nosetests**
```python
nosetests -v thread_exit.py

E
======================================================================
ERROR: thread_exit.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vader/.local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/vader/src/TRIBLER/tribler_ichorid/thread_exit.py", line 42, in test
    main(sys.argv)
  File "/home/vader/src/TRIBLER/tribler_ichorid/thread_exit.py", line 38, in main
    sys.exit(0)
SystemExit: 0

----------------------------------------------------------------------
Ran 1 test in 0.014s

FAILED (errors=1)
Exception in thread Thread-0000000015 (most likely raised during interpreter shutdown):
fish: “nosetests  ./thread_exit.py” terminated by signal SIGSEGV (Address boundary error)
```

**trial**
```python
trial ./thread_exit.py

-------------------------------------------------------------------------------

PASSED
```

**Comment by ichorid** [comment]
This PR is on hold until we finish https://github.com/Tribler/tribler/issues/3799

**Comment by ichorid** [comment]
retest this please

**Comment by ichorid** [comment]
retest this please

**Comment by devos50** [comment]
Hmm, it seems that the `sqlite` library on Mac does not have FTS5 installed.

**Comment by ichorid** [comment]
I've updated both homebrew sqlite and pip sqlite modules. It is either not there, or requrires reboot/library cache reload.

**Comment by devos50** [comment]
IIRC, the SQLite libraries are part of the Python 2.7 installation, which always have priority when importing `sqlite`. Updating sqlite with pip/brew does nothing then.

**Comment by ichorid** [comment]
Moved to #3923

**Pull request closed without merging**